Итерация 9. Записать и прочитать хранилище из файла

thefrol opened the pull request:
Теперь сервер оборачивает хранилище, и умеет перехватывать вызовы с изменением значения в хранилище. Если необходимо пишем в файл

## Сервер
- Теперь завершается нежно:
    + 10 секунд чтобы закрыть запросы
    + Сохраняет данные в файл при выходе
- Умеет сохранять в файл
- Добавлены новые флаги
    + `StoreInterval` - интервал сохранения в файл
    + `Restore` - загружать ли из файла
    + `FileStoragePath` - путь к файлу для сохранения
- Умеет писать синхронно(`StoreInterval` = 0) с запросом, или раз в `StoreInterval` 

## interna/store `MemStorage`
+ Теперь сохраняет себя в файл, или создается из файла
+ Для записи используется формат`gob`

А так же уточнения по логгеру и документации

Commit:
feat: internal/stats Сохранять случайную величину вместе с остальными в одной функции

Commit:
fix: internal/stats переименовать фукнции в благозвучные для пакета

Commit:
fix: cmd/agent объединить env.go и flag.go = config.go

Commit:
fix: cmd/agent Обединить флаги и переменные окружения в однку логику
под руговодством структуры struct

Commit:
fix: cmd/agent выделить работу со счетчиком в internal/stats

Commit:
fix internal/stats выделить описание и константы в отдельный файл internal/stats

Commit:
feat: internal/stats Убрать функции работы со счетчиками, упростить интерфейс пакета

Commit:
fix: cmd/agent исправить поведение агента при настройке(config)

Commit:
fix: cmd/agent Сделать настройки по умолчанию структурой

Commit:
docs: cmd/agent Разметить структуру конфига под sflags

Commit:
fix: cmd/agent исправить порядок функций в config.go, для ясности

Commit:
fix: cmd/server Упаковать конфигурацию в configure(default)

по аналогии с агентом,
 собрать из файлов flag.go и env.go

Commit:
docs: cmd/... Уточнить документацию

Commit:
fix: cmd/server Вернуть забытый flag.Parse()

Commit:
fix: cmd/server сделать конфиг по умолчанию структурой
по аналогии с агентом

Commit:
fix: cmd/agent Использовать httptest в тестах отпраки  вместо самописного

Commit:
fix: cmd/agent Убрать в тесте непроверяемый пример

Commit:
fix: cmd/agent в тестах исправлен нейминг у вспомогательных структур и функций

Commit:
fix: internal/storage Убрать вложенность интерфейсов

Commit:
fix: cmd/agent выделить систему отправки метрик в internal/report

Commit:
fix: cmd/agent Оставить счетчик опросов, Есль отправка с ошибками
не сбрасывать

Commit:
feat: internal/logger Добавить пакет ololog журналирования с базовой функциональностью

Commit:
fix: cmd/server заменить базовое журналирование на ололог

Commit:
feat: Добавить логгирующую мидлварь в сервер
позволяет получать ответы о статусах запросов и ответов

Commit:
feat: /... добавить скрипт для тестирования шестой итерации

Commit:
fix: internal/olologger Отполировать вывод логгера в консоль

Commit:
fix: cmd/server убрать лишнее логгирование из хендлеров

Commit:
fix: cmd/server Проверять Контент-Тайп средствами chi

Commit:
fix: cmd/server Убрать лишний вывод в журнал из хендлеров

Commit:
fix: internal/metrica исправить форматирование

Commit:
feat: internal/metrica добавить структуру для передачи по http

Commit:
feat: добавить (де)сериализаторы easyjson для новой структуры

Commit:
docs: internal/storage добавить соображения по развитию хранилища и метрик

Commit:
feat: internal/metrica Преобразовывать счетчики к Metrica

Commit:
docs: Добавлено новое соображения в РИДМИ

Commit:
fix: /* Заменить "counter" и "gauge" на константы
internal/metrica/const.go

Commit:
feat: internal/metrica Добавить установку значений из типа Metrica

Commit:
feat: cmd/server Добавить POST маршрут /update/
реализовать хендлер который будет обновлять данные по этому маршруту и возвращать новое значение

Commit:
feat: cmd/server Отправлять метрики с маршрута /value/
реализрвать хендлеры, возвращающие данные в формате JSON
используется структура metrica.Metrica

Commit:
fix: internal/stats  добавить потерянную метрику HeapReleased

Commit:
feat: scripts добавить скрипт тестирования седьмой итеарции

Commit:
fix: internal/report переименовать файлы устаревшего протокола

Commit:
feat: internal/storage Добавить новый метод ко хранилищу

Commit:
feat: internal/report добавить способ отправки при помощи POST+JSON

Commit:
fix: cmd/agent переехать на новый api
отправлять данные при помощи POST+JSON

Commit:
fix: internal/report Убрать зависимость отпрки от архитектуры хранилища

Commit:
fix: internal/report Сохранять тело полученного запроса в testHandler

Commit:
feat: internal/report добавить тестирование функции отправки Send()

Commit:
fix: internal/report Дополнить тест проверкой на ошибку

Commit:
docs: коммендарии и документация cmd/server internal/stats+report

Commit:
test: cmd/server добавить тесты по новым маршрутам

Commit:
feat: server+metrica Валидировать получаемый JSON

Commit:
fix: cmd/server Переименовать переменную в стиле Го

Commit:
fix: cmd/repost переименовать переменную в стиле GO

Commit:
fix: cmd/server выделить middleware в отдельный пакет

Commit:
feat: cmd/server Разархивировать запросы в gzip при помощи мидлвари

Commit:
feat: server Логировать, если запрос gzipнут

Commit:
fix: agent Маршалить через easyjson и добавить контракт

Тело запроса должно быть в формате io.Reader, чтобы его могли использовать мидлвари

Commit:
feat: report Ужимать в gzip мидварью
добавлена мидлварь и способ установки, интерфейс не нарушен

Commit:
test: server добавить тест для списка метрик в корневом "/"

Commit:
fix: report выделить мидлвари в отдельный файл

Commit:
feat: report Компрессировать только длинные сообщения
начиная с minLenght, которую можно указать в applyGZIP

Commit:
feat: server Архивировать ответы при помощи мидлвари GZIP

Commit:
fix: server усточнить код по автотестам
+ на странице / теперь возвращаем в контент тайпе text/html
+ вернул отключенные мидлвари

Commit:
fix: server конструктор для CompressedWriter
это решает проблему с кодом по умолчанию

Commit:
fix: server/middleware Выделить функцию ConfirmStatusCode
Очень часто повторяется один и тот же код, собсно, и мы выделаем его в отдельную функцию. Хотя мне уже кажется, что она нас путает и надо ее убрать с учетом последних изменений

Commit:
feat: server/middleware лог тоже проверят на запись кода 0 теперь

Commit:
fix: server Сжимать данные от одного байта, чтобы пустые запросы не требовалось сжимать)
будут 10 байт длинной вместо нуля

Commit:
fix: server Убрать буфер в ListMetrics - по маршруту /

Commit:
fix: server Выводить html страницу по маршруту /
со списком метрик текущей
использовать text/template

Commit:
feat: storage Открыть доступ к мапам счетчиков и гаужей

Commit:
feat: storage Сохранять и читать из файла хранилище (gob)

Commit:
feat: server/router Сохранять читать файлы по http
/save/{file} /load/{file} маршруты

Commit:
fix: storage Разрешить перезаписывать файл с хранилищем по умолчанию

Commit:
fix: middleware.logger Уточнить в логах что именно скомпрессировано, запрос или ответ

Commit:
feat: server Добавить новые настройки
Интервал сохранения,
путь сохранения хранилища
и загружаться ли при старте из хранилища

Commit:
fix: server Подготовить к использованию с закрытием

Commit:
feat: server Обернуть хранилище записывателем сихнронным и нет

Commit:
feat: server Загружать и сохранять сетрики в файл
сделана конфигурация, когда сервер сохраняет в файл, когда загружает из файла при старте, а когда поностью игнорирует сохранение в файлы

Commit:
fix: server Выделить логику инициализации хранилища в отдельный файл

Commit:
feat: server Завершаться в штатном режиме "нежно"
с сохранением в файл

Commit:
fix: cmd/server Проверять, что файл с хранилищем существует

Commit:
feat: script добавить скрипт тестирования 9 итерации

Commit:
fix: server Исправить функцию FileExist
Она проверяла существование файла неправильно, это неправильно запускало сервер, он думал что файл не существует и не читал хранилище из файла

Commit:
fix: server удалить бекдорныпе маршруты /save/{file} и /load/{file}

Commit:
fix: server исправить тест хендлеров

Commit:
fix: server Избавиться от утечки контекста при закрытии сервера

Commit:
fix: storage Изменить имя ресивера, чтобы пройти линтер

Commit:
fix: server middleware разбить gzip по нескольким файлам

Commit:
docs: server gzip Добавить документацию мидлвари

Commit:
feat: gzip server Добавить Apply() опцефункциям

Commit:
feat: gzip server Добавить настройки по умолчанию

Commit:
fix: server уточенение имен, документации. Легкая уборка

Commit:
docs: scheduler документация функций и планы

Commit:
fix: server ungzip Проверять Content-Encoding в правильной последовательсноти

Commit:
fix: agent+fetch отвязаться от хранилища, поправить документацию агента

Теперь при каждом опросе статистики использования памяти не конвертируются метрики, они просто запоминаются.

А перед самой отправкой их можно преобразовать в массив metrica.Metrica при помощи Stats.ToTransport()
лишнее преобразование информации и лишняя связка убрана

Commit:
docs: добавить тудушку

Commit:
fix: fetch Исправить описки в константах

Commit:
feat: Объекдинить пакеты stats и fetch
Удалить старую логику отправки
спрятать логику работы PollCount

Commit:
fix: Выделить gzip сжатие из пакета report в агент

Commit:
feat: выделить slices и scheduler из internal

Commit:
test: Добавить тесты для конфигурирования агента и сервера

Commit:
fix: исправить ошибки конфигурации сервера

МОжно указать пустую строку в переменной окружения теперь

Commit:
docs: server улучшить документацию по конфигу сервера

Commit:
feat: server/config научиться пользоваться -r=false в тесте

Commit:
fix: server/config Поменять тип временного интервала на uint
теперь можно только положительные значения давать

Commit:
fix: agent/config Поменять тип временного интервала на uint
теперь можно только положительные значения давать

EsdeathEsdeath commented on cmd/agent/agent.go:
```diff
@@ -14,40 +16,41 @@ var store storage.Storager
+	c.AddJob(time.Duration(config.ReportInterval)*time.Second, func() {
 		//отправляем на сервер
-		err := sendStorageMetrics(store, "http://"+*addr)
+		err := report.Send(store.Metricas(), "http://"+path.Join(config.Addr, "update"))
```
Через `Sprintf` конкатенация строк будет эффективнее, чем через `+`

thefrol replied:
реализовано в [1c9bc53](https://github.com/thefrol/kysh-kysh-meow/pull/18/commits/1c9bc53d4a323a905245860b36234ea7c2da7bae)

EsdeathEsdeath commented on cmd/agent/config.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+type config struct {
+	Addr            string `env:"ADDRESS" flag:"~a" desc:"(строка) адрес сервера в формате host:port"`
```
Чтоб не хранить дефолтные значения в отдельными переменными, можно их добавить сюда через `env-default`

thefrol replied:
Я бы очень хотел так сделать, но к сожалению так невозможно ;< Вот строчка из официальной документации 

```
Default values defined as struct tags will overwrite existing values during Parse.
```

Что означает, что `envDefault:"localhost:8080"` обязательно перепишет, то, что мы указали в параметрах командной строки ранее, что печально, поэтому мне и приходится создавать эту структуру с дефолтными значениями. Единственное, я думаю, эту структуру можно из мейна сюда перенести

thefrol replied:
Если на примере:

агент запускается командой ` ./cmd/server/server -p 3`

```go

	flag.UintVar(&cfg.PollingInterval, "p", 1, "число, частота опроса метрик")

	flag.Parse()

	ololog.Debug().Uint("polling interval", cfg.PollingInterval)
        // DBG polling interval 3

	err := env.Parse(&cfg)
	if err != nil {
		panic(err)
	}

	ololog.Debug().Uint("polling interval", cfg.PollingInterval).Msgf("%v", cfg.PollingInterval)
        // DBG polling interval 1
```

Итого дефолтные значения переписали ввод с командной строки.

Я вообще ищу какой-нить пакет, который позволял бы и командную строку, и переменные окружения не кринжово сделать, но пока никак. Нашел один, но там какие-то проблемы с `go.mod` и типа это надо себе форкнуть поправить, а у автора видит пулл-реквест с 21 года

thefrol replied:
Реализовано в [d9e4e24](https://github.com/thefrol/kysh-kysh-meow/pull/18/commits/d9e4e24b557fb40029a9d1442d1001fb570d28c8)

Оставил сигнатуру 

```go
cfg := mustConfigure(defaultConfig)
```

так можно кликнуть в ИДЕ или гитхабе и поправить если что  быстро, и тестируется хорошо

EsdeathEsdeath replied:
Все окей, про `env` -дельное замечание, так как флаги командной строки используют не очень часто, вот и не делают не кринжовый вариант, не пользуется спросом, кстати это к сегодняшней встречи, если хочется пописать что-то еще, почему бы не попробовать реализовать собственную либу?

EsdeathEsdeath commented on cmd/agent/config.go:
```diff
@@ -0,0 +1,43 @@
+
+func init() {
+	// добавляет смайлик кота в конец справки
+	flag.Usage = func() {
```
Тут же можно инициализировать логгер и везде использовать в агенте, чтоб уйти от принтов

thefrol replied:
Просто я даже не знаю, как тут инициализировать логгер, ибо он у меня инициализируется прям в пакете `ololog`. Но если дропну этот пакет за ненадобностью, тогда да поставлю сюда идея хорошая

И вроде принтов-то и нет. Я специально посмотреть через поиск. Тут используется папа во flag.Usage, это просто чтобы сформировать правильное сообщение для пакета `flag`, чисто визуально тут логгер не нужен. Есть ещё пара использований Sprint(), но не в консоль.

EsdeathEsdeath replied:
Но я не вижу большой необходимости колбасить свой логер

EsdeathEsdeath commented on cmd/server/config.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/caarlos0/env/v6"
+)
+
+type config struct {
```
Тут те же комментарии

EsdeathEsdeath commented on cmd/server/handlers.go:
```diff
@@ -2,15 +2,129 @@ package main
 
 import (
 	"fmt"
```
А вот это уже хорошо бы убрать в `internal`

thefrol replied:
Сделано вот в этом ПР https://github.com/thefrol/kysh-kysh-meow/pull/20

EsdeathEsdeath commented on cmd/server/handlers.go:
```diff
@@ -2,15 +2,129 @@ package main
+		return
+	}
+
+	// returning можно сделать кой-то отдельной функицей
```
В принципе вообще все? что ты используешь 2+ раза, можно выносить отдельно, чтоб не плодить массу лишнего кода

thefrol replied:
Там если разобраться, то у меня не получилось повторно использовать, вот в этом коммите [f45d05a](https://github.com/thefrol/kysh-kysh-meow/pull/20/commits/f45d05aad347260cdb2fc028004eea365150669a)

Commit:
feat: config Теперь при неправильных, отприцательных значениях паниковать
добавить тесты, которые это все проверяют

Вообще, конечно, это тесты показали, что немного не до конца валидируются значения

EsdeathEsdeath commented on cmd/server/handlers.go:
```diff
@@ -2,15 +2,129 @@ package main
+		return
+	}
+
+	if !found {
```
Вот этот кусок выглядит так, что его можно вынести и переиспользовать

thefrol replied:
пока короче разобрался с повторным использованием, все же не смог их повторно использовать сообажения в этом коммите [f45d05a](https://github.com/thefrol/kysh-kysh-meow/pull/20/commits/f45d05aad347260cdb2fc028004eea365150669a)

EsdeathEsdeath commented on cmd/server/handlers.go:
```diff
@@ -88,31 +194,43 @@ func getValue(w http.ResponseWriter, r *http.Request, params URLParams) {
 
-	cl := store.ListCounters()
-	gl := store.ListGauges()
+	t := `
```
Если переменная используется часто, то лучше ее вообще вынести в константу или просто из функции, а то каждый раз при вызове мы ее инициализируем

thefrol replied:
Спасибо, хорошее замечание, я даже совсем об этом не подумал, что это так будет работать. К счастью, сейчас этой переменной уже там нет.

EsdeathEsdeath commented on cmd/server/middleware/log.go:
```diff
@@ -0,0 +1,106 @@
+			// запрос такой-то
+			// тут сообщения от хендера посередине
+			// ответ такой-то
+			// Чтобы сообщения были как бы обернуты миддлеваром, вот начало вот конец.
```
Боюсь, что так будет сделать проблематично, да и на практике, обычно, такое не нужно, проблему, что ты описал скорее решает трассировка, которая и проследит весь путь внутри программы и извне, если интересно посмотри в ту сторону, в качестве примера можешь глянуть на `OpenTelemetry`

thefrol replied:
Да, справедливо. Опять же. Если несколько запросов одновременно - без трассировки все перепутается.

EsdeathEsdeath commented on cmd/server/router.go:
```diff
@@ -4,17 +4,49 @@ import (
+		middleware.MinLenght(1),
+		// todo
+		//
+		// Хочу чтобы это выглядело так compress.ContentType(...)
```
Посмотри точно ли `chi middleware.Compress` не решит проблему, с учетом того, что насколько помню там есть список того, что он сжимает.
Либо я не очень понял проблему

thefrol replied:
Да `chi.Compress()` все решает, кроме минимальной длинны. Просто хотелось свои силы попробовать в компрессии, в качестве учебной штуки. А тут комментарии скорей о том, как я бы хотел, чтобы это выглядело

Commit:
docs: добавлен инсайт в ридми

thefrol committed:
Merge pull request #19 from thefrol/iter9.1/refactor

Рефакторинг 9.1 После второго спринта https://github.com/thefrol/kysh-kysh-meow/pull/19

EsdeathEsdeath commented on cmd/server/server.go:
```diff
@@ -4,31 +4,92 @@
+	// Listen for syscall signals for process to interrupt/quit
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+	go func() {
```
Завершения горутины тоже бы дождаться, но это будет дальше в спринтах

thefrol replied:
Имеется в виду sync.WaitGroup?

thefrol replied:
пока оставил до лучших времен, ну и поставил себе в TODO

EsdeathEsdeath commented on cmd/server/server.go:
```diff
@@ -4,31 +4,92 @@
+	// saver позволяет получить доступ к функции сохранения в хранилище,
+	// потому что в Storager мы его не имеем, и не очень-то хотим
+	if cfg.FileStoragePath != "" { //todo можно сделать как функцию в структуре config
+		type saver interface {
```
А вот так лучше не делать конечно

EsdeathEsdeath commented on cmd/server/storage.go:
```diff
@@ -0,0 +1,97 @@
+package main
```
Тоже луче вынести 
Вообще в cmd скопилось много всего, удобнее оставить там максимум `main config`, а остальное убрать

thefrol replied:
Сделано вот в этом ПР https://github.com/thefrol/kysh-kysh-meow/pull/21 - стало гораздо быстрее работать лол

Commit:
fix: server переименовать функцию по совету go ver

EsdeathEsdeath commented on internal/ololog/docs.go:
```diff
@@ -0,0 +1,32 @@
+// Это не сложно, но надо ли оно мне сейчас. Пока пусть работает как есть
+//
+//
+// Блин, надо было пакет назвал Оло, или олло, был бы оло.лог(), оло.л()
```
😆

EsdeathEsdeath commented on internal/ololog/log.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/rs/zerolog"
+)
+
+func Log() *zerolog.Event {
```
А зачем свое? доггер на проекте меняется крайне редко или вообще никогда

thefrol replied:
Короче по итогу `олологгер` я просто убрал в [bb37574](https://github.com/thefrol/kysh-kysh-meow/pull/18/commits/bb375746cc7448a1925e6563ea1ecb363267e4c1) он как бы польностью дублировал `zerolog`. А я его так много писал в коде, что мне уже не стало больше смешно достаточно) Так что помянем. 

Ну плюс названия компонентов должны как бы намекать, что они делают, а `ололог` скорее вводил в заблуждения и может даже вызывал брезгливость

EsdeathEsdeath commented on scripts/test-iter9:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
А вот это все лучше оставить локально, тут оно выглядит излишним

thefrol replied:
Удалил вот тут [75c7b15](https://github.com/thefrol/kysh-kysh-meow/pull/18/commits/75c7b15dcda5539c3c6fc78cd1538ed51f16fd8f)

Хотелось конечно, чтобы я мог сесть за другую машину и там работать. Но линукс у меня только тут. Так что может не так страшно. 

Зато изучил прекрасную команду `git rm --cached`

Commit:
fix: agent Соединять путь через Sprintf, добавить тесты

Commit:
fix: перенести конфиг по умолчанию в файл config.go

Commit:
fix: server хендлеры, мидлвари и роутер убраны в отдельный пакет
поправлены тесты
тесты проходят, свои и от практикума

Commit:
fix: server/handlers разделить хендлеры по файлам
один для старого апи, один для нового

Commit:
fix: handlers старые хендлеры теперь apiv1, новые apiv2

Commit:
fix: cmd/server отказаться от глобальной переменной store

Commit:
fix: report добавлена обработка пропущенной ошибки

Commit:
fix: server log middleware Правильно выводить код ответа

Commit:
test: router перекинуть в тесты в один файл

Commit:
fix: почистить MeowRouter() для сервера
добавить документацию
убрать лишние комменты
сгруппировать маршруты
убрать корявое форматирование

Commit:
fix: почистить apiV1 сервера
+ убрать бесячую обертку AsMetricURL - теперь это просто функция, которая запускается внутри хендлера,
+ документировать функции
+ Переименовать хендлер выдающий ошибку прям в ошибку, чтобы было голословно

Commit:
docs: Добавить документацию пакетам apiv1 apiv2

Commit:
fix: переименовать файл в агенте с компрессией

Commit:
fix: internal/metrica оптимизировать валидацию Metrica

Commit:
fix: server переработать json хендлеры
+ теперь работают быстрее,
+ логику проверок и работы с хранилищем вывести в отдельные функции
+ дополнить документацию
+ модифицировать объект в запросе(стремно но ради скорости)

Commit:
fix: apiV2 Сделать информативней сообщение об ошибке

Commit:
fix: apiv1 избавиться от глобальной переменной store
теперь апи это структура API c конструктором new
и своим личным интерфейсом для хранилища, который нужен именно этому API

Commit:
fix: apiv1 исправиль описание ошибки

Commit:
fix: рефорк apiv2 с отвязыванием от глобального хранилища

Commit:
docs: Добавить инсайт по итогам пул-ревеста

thefrol committed:
Merge pull request #20 from thefrol/iter9.2/moving-server-files

9.2 Рефактор хендлеров

thefrol commented:
Проведен рефакторинг хендлеров жесткий https://github.com/thefrol/kysh-kysh-meow/pull/20

## Обновление 9.2

Теперь `MemStorage` зависит от хендлеров, об этом даже думать приятно. Что хендлеры у меня будут в порядке вне зависимости от того, чтобы я там с ними не делал. Все требования у меня тут под рукой прямо. В т `же время в  `MemStorage` добавлены проверки на соответствие `apiv1.Storager`

Это все произошло потом, что я выделил в отдельные пакеты все хендлеры и там с ними разобрался. Иначе я не мог бы сделать зависимость хендлерам от функции пакета main сервера. Потому что бы получалась циклическая зависимости. **Действительно очень полезно складывать все в `internal`**

Именно благодаря этому я могу перенести экспортные функции в пакет и не париться о них. Потому что `MemStorage` - зависимая штука

Основные коммиты: 
+ Хендлеры разбиты на два пакета `apiv1` и `apiv2`
+ избавиться от глобальной переменной `store` в `apiv1`. Теперь апи это структура `API` c конструктором `New`
и своим личным интерфейсом для хранилища, который нужен именно этому API
+ аналогично `apiv2`
+ `apiv2` жестко переработан в плане скорости
+ мидлвари выделены в internal
+ роутер выделен в пакет `router` папки `internal`

По этогам этого рефакторинга вот что хочется сказать 

> **Выделяй пакеты из программ. ** Это очень помогает, когда хочешь настроить зависимости одних классов или компонентов от других. Пока все лежит в пакете `main`, мы даже на уровне кода не можем прописать зависимости. Мы не можем ссылаться на пакет `main` из других пакетов. А когда такая возможность появляется, голова как бы сразу проясняется. Да, это порой бывает непросто. Тут же вознимают проблемы с глобальными переменными и прочим, но но того стоит

Этот инсайт помещен в документацию

Commit:
fix: storage Добавить  на соответсвие интерфейса apiv2

Commit:
feat: удалить ololog

Commit:
fix: удалить скрипты для тестирования из репозитория

Commit:
feat: storage Изименить интерфейс сохранения хранилища,
Теперь при загрузке из файла - то, что в памяти перезаписывается

Commit:
docs: Добавить тудушку-идею в РИДМИ

Commit:
feat: Создать новые классы хранилищ с сохранением
+ IntervalicDump сохраняет с интервалом
+ SyncDump - сохраняет в момент изменения метрики
+ переписать сервер на новые классы

Commit:
fix: удалить старые обертки для хранилища из сервера

Commit:
fix: storage убрать неиспользуюмую переменную из сигнатуры

Commit:
docs: storage задокументировать новые хранилища классы

Commit:
fix: storage исправить текст ошибки

Commit:
feat: storage Вынести запись и чтение из memStore в FileStorage

Commit:
fix: cmd/server вынести запуск сервера в отдельную функцию

Commit:
fix: storage Вынести FileStorage в отдельный файл

Commit:
fix: storage переименовать файл с исходником

Commit:
fix:  server документация и дополнительный логгинг

thefrol committed:
Merge pull request #21 from thefrol/iter9.3/storage-decomposition

Итерация 9.3 Переработка хранилища

thefrol commented:
## Обновление 9.3
Удивительно, но сервер стал работать раза в три быстрее. Интересно, с чего вдруг

+ Удалены старые велосипеды с записью в хранилища, они переработаны в отдельные классы `IntervalicDump` и `SyncDump`. 
+ При этом `Memstore`  лишился возможности записи на диск, для этого теперь используется его "наследник" `FileStorage`
+ В сервере больше нет кода, создающего хранилища
+ Новые классы задокументированы
+ Почищен и задокументирован код сервера

thefrol commented:
Удален `ололог`  [bb37574](https://github.com/thefrol/kysh-kysh-meow/pull/18/commits/bb375746cc7448a1925e6563ea1ecb363267e4c1) 😢

thefrol commented:
Вообще, я понял, что я как-то совсем невнимательно читал [про папки в ГО](https://github.com/golang-standards/project-layout), потому что там конечно написано, что в `cmd` должно лежать по минимуму, и как организовать код приложения в папке `internal`

EsdeathEsdeath commented on cmd/agent/compress.go:
```diff
@@ -1,4 +1,4 @@
-package report
```
Это, все же не должно лежать в `main`

EsdeathEsdeath commented on cmd/agent/config.go:
```diff
@@ -3,37 +3,55 @@ package main
-	env.Parse(&cfg)
+	err := env.Parse(&cfg)
+	if err != nil {
+		panic(err)
```
Вот такое избегаем, хотя бы `log.Fatal`

EsdeathEsdeath commented on cmd/server/config.go:
```diff
@@ -3,32 +3,66 @@ package main
-	env.Parse(&cfg)
+	err := env.Parse(&cfg)
+	if err != nil {
+		panic(err)
```
И тут, вообще `panic` в коде это моветон

thefrol replied:
сделано https://github.com/thefrol/kysh-kysh-meow/pull/37

EsdeathEsdeath commented on cmd/server/server.go:
Все же `cmd/server/main.go`

EsdeathEsdeath commented on cmd/server/server.go:
```diff
@@ -1,41 +1,49 @@
+
+	// Завершаем последние дела
+	// попытаемся сохраниться в файл
+	cancelStorage()
```
Надеюсь через `defer`?

EsdeathEsdeath commented on cmd/server/server.go:
```diff
@@ -45,51 +53,93 @@ func main() {
+//
+// На входе получает экземпляр хранилища m, и далее оборачивает его другим классов,
+// наиболее соответсвующим задаче, исходя из cfg
+func ConfigureStorage(m *storage.MemStore, cfg config) (storage.Storager, context.CancelFunc) {
```
Это лучше отсюда убрать хотя бы в условный `cmd/server/init.go`, `main` должен быть максимально компактным

EsdeathEsdeath commented on internal/report/pollcount.go:
```diff
@@ -0,0 +1,16 @@
+// DropPoll сбрасывает значение счетчика опросов памяти в указанном хранилище
+// dropCounter сбраcывает счетчик
+func dropPollCount() {
+	pollCount = 0 // todo поллкаунт может стать вообще внутренней штукой модуля report если их объдинить, тогда агенту вообще не надо будет об этом париться
```
Верное направление, ему и не надо

thefrol replied:
Сделано где-то в апдейтах, но точно так сходу не найду

EsdeathEsdeath commented on internal/server/api/v1/handlers.go:
```diff
@@ -0,0 +1,167 @@
+
+// Storager это интерфейс к хранилищу, которое использует именно этот API. Таким образом мы делаем хранилище зависимым от
+// API,  а не наоборот
+type Storager interface {
```
Лучше бы все-таки какие то ошибки возвращать и предусмотреть это на уровнях выше, не хочется панику где-нибудь словить и потом гадать что пошло не так

thefrol replied:
Да справедливо, добавлено в https://github.com/thefrol/kysh-kysh-meow/pull/23

EsdeathEsdeath commented on internal/server/api/v1/handlers.go:
```diff
@@ -0,0 +1,167 @@
+type Storager interface {
+	Counter(name string) (metrica.Counter, bool)
+	SetCounter(string, metrica.Counter)
+	ListCounters() []string
```
А если у нас и `ListCounters()` и  `ListGauges()` имеют одинаковую сигнатуру, то может их объединить?

thefrol replied:
блин, классная идея!

thefrol replied:
https://github.com/thefrol/kysh-kysh-meow/issues/26

EsdeathEsdeath commented on internal/server/api/v1/handlers.go:
```diff
@@ -0,0 +1,167 @@
+
+// API это колленция http.HanlderFunc, которые обращаются к единому хранилищу store
+type API struct {
+	store Storager
```
А логер не хотим тут?

thefrol replied:
Он у меня просто глобальный теперь, да и был в общем) https://github.com/thefrol/kysh-kysh-meow/pull/23

thefrol replied:
Или не стоит так делать

EsdeathEsdeath commented on internal/server/api/v1/handlers.go:
```diff
@@ -0,0 +1,167 @@
+
+// New создает новую
+func New(store Storager) API {
+	if store == nil {
```
По хорошему такое условие вообще не должно выполняться, на уровне main мы идем снизу вверх, если хранилища нет - то и дальше мы можем не работать

thefrol replied:
Да справедливо. Код по сути ерундой занимается. Все эти апишки бессмысленные были удалены в https://github.com/thefrol/kysh-kysh-meow/pull/23

EsdeathEsdeath commented on internal/server/api/v1/handlers.go:
```diff
@@ -0,0 +1,167 @@
+
+	value, err := strconv.ParseInt(params.value, 10, 64)
+	if err != nil {
+		w.Header().Add("Content-Type", "text/plain")
```
Если мы часто используем что-то такое:
```
		w.Header().Add("Content-Type", "text/plain")
 		w.WriteHeader(http.StatusBadRequest)
 		http.Error(w, "^0^ Ошибка значения, не могу пропарсить", http.StatusBadRequest)
```
 Может унифицируем и вынесем? Зачем плодить код, помним про DRY

thefrol replied:
Сделал в https://github.com/thefrol/kysh-kysh-meow/pull/23

EsdeathEsdeath commented on internal/server/api/v2/json_handlers.go:
```diff
@@ -0,0 +1,219 @@
+
+// Storager это интерфейс к хранилищу, которое использует именно этот API. Таким образом мы делаем хранилище зависимым от
+// API,  а не наоборот
+type Storager interface {
```
Ты же его уже объявлял? Может вынести на уровне пакета в отдельный файл и переиспользовать, чуть сократим количество кода

thefrol replied:
Тоже сделано в https://github.com/thefrol/kysh-kysh-meow/pull/23

EsdeathEsdeath commented on internal/server/api/v2/json_handlers.go:
```diff
@@ -0,0 +1,219 @@
+// Этот пакет содержит хендлеры нового образца,
+// где мы передаем значения при помощи json-запросов
+// по маршрутам /update и /value
+package apiv2
```
А точно ли `v2`? Обычно мы используем версионирование и, в частности, `v2`, когда у нас изменения ломающие обратную совместимость, а тут по сути ты же просто расширяешь API?

EsdeathEsdeath commented on internal/server/router/router.go:
```diff
@@ -0,0 +1,97 @@
+		//r.With(chimiddleware.AllowContentType("text/plain"))
+		r.Get("/value/{type}/{name}", v1.GetValue)
+
+		r.Post("/update/{type:counter}/{name}/{value}", v1.UpdateCounter)
```
Можно объединить на уровне саброутеров, станет чуть компактнее

thefrol replied:
Вот конкретно два этих маршрута не совсем понимаю как объединить. Мне кажется после обновки https://github.com/thefrol/kysh-kysh-meow/pull/23 уже нормально смотрится

EsdeathEsdeath commented on internal/server/router/router.go:
```diff
@@ -0,0 +1,97 @@
+// Испрользуется так
+// router := chi.NewRouter()
+// InstallAPIV2(router, store)
+func InstallAPIV2(r chi.Router, v2 apiv2.API) {
```
Все еще не понимаю, если ломающие изменения, то почему вообще все не перенести на v2 и не работать с ним, тем более, что в таком случае методы должны стать `/v2/update`...

thefrol replied:
Да согласен, контринтуитивное наименование. В итоге я все апи слил обратно, когда они стали похожим образом работать после вот этого обновления https://github.com/thefrol/kysh-kysh-meow/pull/23

EsdeathEsdeath commented on internal/storage/filestorage.go:
```diff
@@ -0,0 +1,89 @@
+	if !fileExist(s.FileName) {
+		return ErrorRestoreFileNotExist
+	}
+	file, err := os.Open(s.FileName)
```
Где-то же закрываем? Да?

thefrol replied:
Исправлено вот тут [ce9673e](https://github.com/thefrol/kysh-kysh-meow/pull/25/commits/ce9673e11435203672268352c30368190ae22005)

Спасибо, что заметил!
